### PR TITLE
Add overflow tests

### LINKED ISSUE #541: Add tests for integer overflow
The application performs a calculation that can produce an integer overflow or underflow. This can happen when an application's logic assumes that a resulting value will always be larger or smaller than the original value. This can lead to other weaknesses because the size of the buffer is incorrectly calculated, such as a buffer overflow.

Use the following guidelines to test for integer overflow:
- Identify sections of the site that perform integer arithmetic with user-supplied values.
- Provide large negative and positive values and inspect the result.
- This test fails if the result is negative when it should be positive, or vice-versa.

### PR DESCRIPTION
**Pull request by PatKamin**
### Description

closes #541

### Checklist

- [ ] Code compiles without errors locally
- [ ] All tests pass locally
- [ ] CI workflows execute properly

- [ ] CI workflows, not executed per PR (e.g. Nightly), execute properly 
- [ ] New tests added, especially if they will fail without my changes
- [ ] Added/extended example(s) to cover this functionality
- [ ] Extended the README/documentation
- [ ] All newly added source files have a license
- [ ] All newly added source files are referenced in CMake files
- [ ] Logger (with debug/info/... messages) is used

### CODE REVIEW

File: test/provider_os_memory_multiple_numa_nodes.cpp [hunk 1]
```diff
@@ -711,3 +711,21 @@ TEST_F(testNuma, checkModeInterleaveIllegalArgSet) {
...
+    os_memory_provider_params.numa_mode = UMF_NUMA_MODE_INTERLEAVE;
+    os_memory_provider_params.part_size = SIZE_MAX;
+    os_memory_provider_params.numa_list = numa_nodes.data();
+    os_memory_provider_params.numa_list_len = numa_nodes.size();
+
+    auto res = umfMemoryProviderCreate(umfOsMemoryProviderOps(),
+                                       &os_memory_provider_params,
+                                       &os_memory_provider);
+    ASSERT_EQ(res, UMF_RESULT_SUCCESS);
+    ASSERT_EQ(os_memory_provider, nullptr);
```
Comment (lukaszstolarczuk) [inline test/provider_os_memory_multiple_numa_nodes.cpp:730]: so provider is null, but we expect UMF_RES_SUCCESS?

Comment (PatKamin) [reply test/provider_os_memory_multiple_numa_nodes.cpp:730]: Done

**Review by lplewa — CHANGES_REQUESTED** [review]

File: src/provider/provider_os_memory.c [hunk 2]
```diff
@@ -360,6 +360,15 @@ validatePartitions(umf_os_memory_provider_params_t *params) {
     return UMF_RESULT_SUCCESS;
 }
 
+static umf_result_t validatePartSize(os_memory_provider_t *provider, umf_os_memory_provider_params_t *params) {
+    size_t page_size;
+    if (ALIGN_UP(params->part_size, os_get_min_page_size(provider, NULL, &page_size)) < params->part_size) {
```
[the remark this answers is not in the document]
Comment (KFilipek) [reply src/provider/provider_os_memory.c:365]: Write C++ template function align_up_safe<T>(value).

File: src/base_alloc/base_alloc.c [hunk 3]
```diff
@@ -134,7 +134,10 @@ static void *ba_os_alloc_annotated(size_t pool_size) {
 }
 
 umf_ba_pool_t *umf_ba_create(size_t size) {
-    size_t chunk_size = ALIGN_UP(size, MEMORY_ALIGNMENT);
+    size_t chunk_size = ALIGN_UP_SAFE(size, MEMORY_ALIGNMENT);
+    if (!chunk_size) {
```
Comment (ldorau) [inline src/base_alloc/base_alloc.c:138]: IMHO in case of numeric values you should use:
```c
 if (chunk_size == 0)
```

(instead of ` if (!chunk_size)`) - also in all places below.

File: src/utils/utils_common.h [hunk 4]
```diff
@@ -29,8 +29,13 @@ extern "C" {
         expression;                                                            \
     } while (0)
 
-#define ALIGN_UP(value, align) (((value) + (align)-1) & ~((align)-1))
-#define ALIGN_DOWN(value, align) ((value) & ~((align)-1))
+#define ALIGN_UP(value, align) ((value + align - 1) & ~(align - 1))
```
Comment (ldorau) [inline src/utils/utils_common.h:32]: It should stay as was:
```c
#define ALIGN_UP(value, align) (((value) + (align)-1) & ~((align)-1))
```
Please revert it.

File: src/utils/utils_common.h [hunk 5]
```diff
@@ -29,8 +29,13 @@ extern "C" {
...
+#define ALIGN_UP_SAFE(value, alignment)                                        \
+    ((alignment == 0) ? (value)                                                \
+                      : ((value + alignment - 1) < value                       \
+                             ? 0                                               \
+                             : ((value + alignment - 1) & ~(alignment - 1))))
+#define ALIGN_DOWN(value, align) (value & ~(align - 1))
```
Comment (ldorau) [inline src/utils/utils_common.h:38]: It should stay as was:
```c
#define ALIGN_DOWN(value, align) ((value) & ~((align)-1))
```
Please revert it.

File: src/utils/utils_common.h [hunk 6]
```diff
@@ -29,8 +29,13 @@ extern "C" {
         expression;                                                            \
     } while (0)
 
-#define ALIGN_UP(value, align) (((value) + (align)-1) & ~((align)-1))
-#define ALIGN_DOWN(value, align) ((value) & ~((align)-1))
+#define ALIGN_UP(value, align) ((value + align - 1) & ~(align - 1))
+#define ALIGN_UP_SAFE(value, alignment)                                        \
```
Comment (ldorau) [inline src/utils/utils_common.h:33]: All occurrences of `value` and `alignment` should be in parenthesis (`(value)` and `(alignment)`), in case they were expressions.

File: src/provider/provider_os_memory.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (PatKamin) [reply src/provider/provider_os_memory.c]: Using ALIGN_UP_SAFE macro now. Returning 0 on overflow.

File: test/common/provider.hpp [hunk 7]
```diff
@@ -106,6 +106,10 @@ struct provider_malloc : public provider_base_t {
             align = 8;
         }
 
+        if (SIZE_MAX - size < align) {
```
[the remark this answers is not in the document]
Comment (PatKamin) [reply test/common/provider.hpp]: Replaced the check with ALIGN_UP_SAFE macro.

File: src/base_alloc/base_alloc.c (hunk 3 above)
Comment (PatKamin) [reply src/base_alloc/base_alloc.c]: Done

File: src/utils/utils_common.h (hunk 4 above)
Comment (PatKamin) [reply src/utils/utils_common.h]: Reverted

File: src/utils/utils_common.h (hunk 5 above)
Comment (PatKamin) [reply src/utils/utils_common.h]: Reverted

File: src/base_alloc/base_alloc_global.c [hunk 8]
```diff
@@ -155,10 +155,19 @@ void *umf_ba_global_aligned_alloc(size_t size, size_t alignment) {
         return NULL;
     }
 
+    if (SIZE_MAX - size < ALLOC_METADATA_SIZE) {
```
Comment (ldorau) [inline src/base_alloc/base_alloc_global.c:162]: Maybe `(size > SIZE_MAX - ALLOC_METADATA_SIZE)` - it is more obvious

File: src/base_alloc/base_alloc_global.c [hunk 9]
```diff
@@ -155,10 +155,19 @@ void *umf_ba_global_aligned_alloc(size_t size, size_t alignment) {
...
+        LOG_ERR("base_alloc: allocation size (%zu) too large.", size);
+        return NULL;
+    }
+
     // for metadata
     size += ALLOC_METADATA_SIZE;
 
     if (alignment > ALLOC_METADATA_SIZE) {
+        if (SIZE_MAX - size < alignment) {
```
Comment (ldorau) [inline src/base_alloc/base_alloc_global.c:171]: `(size > SIZE_MAX - alignment)` is more clear

File: test/common/provider.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (ldorau) [reply test/common/provider.hpp]: Not replaced yet.

File: src/base_alloc/base_alloc_global.c [hunk 10]
```diff
@@ -96,8 +96,12 @@ static void *add_metadata_and_align(void *ptr, size_t size, size_t alignment) {
     if (alignment <= ALLOC_METADATA_SIZE) {
         user_ptr = (void *)((uintptr_t)ptr + ALLOC_METADATA_SIZE);
     } else {
-        user_ptr =
-            (void *)ALIGN_UP((uintptr_t)ptr + ALLOC_METADATA_SIZE, alignment);
+        user_ptr = (void *)ALIGN_UP_SAFE((uintptr_t)ptr + ALLOC_METADATA_SIZE,
+                                         alignment);
+        if (!user_ptr) {
+            LOG_ERR("base_alloc: ptr alignment overflow");
```
Comment (ldorau) [inline src/base_alloc/base_alloc_global.c:102]: `pointer alignment overflow`

File: src/base_alloc/base_alloc_linear.c [hunk 11]
```diff
@@ -88,7 +88,11 @@ umf_ba_linear_pool_t *umf_ba_linear_create(size_t pool_size) {
         pool_size = MINIMUM_LINEAR_POOL_SIZE;
     }
 
-    pool_size = ALIGN_UP(pool_size, ba_os_get_page_size());
+    pool_size = ALIGN_UP_SAFE(pool_size, ba_os_get_page_size());
+    if (pool_size == 0) {
+        LOG_ERR("umf_ba_linear_create: pool_size page alignment overflow");
```
Comment (ldorau) [inline src/base_alloc/base_alloc_linear.c:93]: `LOG_ERR("umf_ba_linear_create: ` - I am almost sure, that LOG_* functions always prints out the function prefix, so it is not needed.

File: test/common/provider.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (ldorau) [reply test/common/provider.hpp]: Replaced, but in another commit that the original issue was submitted.

File: src/base_alloc/base_alloc_linear.c (hunk 11 above)
Comment (PatKamin) [reply src/base_alloc/base_alloc_linear.c]: Removed prefixes

File: src/base_alloc/base_alloc_global.c (hunk 10 above)
Comment (PatKamin) [reply src/base_alloc/base_alloc_global.c]: Done

File: src/base_alloc/base_alloc_global.c (hunk 9 above)
Comment (PatKamin) [reply src/base_alloc/base_alloc_global.c]: Done

File: src/base_alloc/base_alloc_global.c (hunk 8 above)
Comment (PatKamin) [reply src/base_alloc/base_alloc_global.c]: Done

File: src/base_alloc/base_alloc_global.c [hunk 12]
```diff
@@ -96,8 +96,12 @@ static void *add_metadata_and_align(void *ptr, size_t size, size_t alignment) {
     if (alignment <= ALLOC_METADATA_SIZE) {
         user_ptr = (void *)((uintptr_t)ptr + ALLOC_METADATA_SIZE);
     } else {
-        user_ptr =
-            (void *)ALIGN_UP((uintptr_t)ptr + ALLOC_METADATA_SIZE, alignment);
+        user_ptr = (void *)ALIGN_UP_SAFE((uintptr_t)ptr + ALLOC_METADATA_SIZE,
+                                         alignment);
+        if (!user_ptr) {
+            LOG_ERR("base_alloc: pointer alignment overflow");
```
Comment (lukaszstolarczuk) [inline src/base_alloc/base_alloc_global.c:102]: remove these prefixes? `base_alloc: ` (at least 3x in this file)

File: test/common/provider.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (PatKamin) [reply test/common/provider.hpp]: Right, it should be ok now.

File: src/provider/provider_os_memory.c [hunk 13]
```diff
@@ -341,6 +341,22 @@ validatePartitions(umf_os_memory_provider_params_t *params) {
     return UMF_RESULT_SUCCESS;
 }
 
+static umf_result_t os_get_min_page_size(void *provider, void *ptr,
```
Comment (lukaszstolarczuk) [inline src/provider/provider_os_memory.c:344]: don't we already have a function like this?

File: src/utils/utils_common.h [hunk 14]
```diff
@@ -40,6 +40,13 @@ typedef enum umf_purge_advise_t {
 #define IS_NOT_ALIGNED(value, align)                                           \
     ((align != 0 && (((value) & ((align)-1)) != 0)))
 #define ALIGN_UP(value, align) (((value) + (align)-1) & ~((align)-1))
```
Comment (lukaszstolarczuk) [inline src/utils/utils_common.h:42]: do we still need `ALIGN_UP`? if yes, pls add a comment when to use the safe one

File: src/base_alloc/base_alloc_global.c (hunk 12 above)
Comment (PatKamin) [reply src/base_alloc/base_alloc_global.c:102]: I've used the style of the messages that was already used in this file. Note that these are not referring to the local function's name.

File: src/utils/utils_common.h (hunk 14 above)
Comment (PatKamin) [reply src/utils/utils_common.h:42]: Yes, we still need it for cases where `0` is a valid input, like in the `validatePartSize()` function I'm introducing with this PR.

File: src/provider/provider_os_memory.c (hunk 13 above)
Comment (PatKamin) [reply src/provider/provider_os_memory.c:344]: Yes, I have to declare it before use. Removed the redundant declaration below, though.

File: src/utils/utils_common.h [hunk 15]
```diff
@@ -40,6 +40,13 @@ typedef enum umf_purge_advise_t {
...
+#ifndef ALIGN_UP_SAFE
+#define ALIGN_UP_SAFE(value, align)                                            \
+    (((align) == 0) ? (value)                                                  \
+                    : (((value) + (align)-1) < (value)                         \
+                           ? 0                                                 \
+                           : (((value) + (align)-1) & ~((align)-1))))
```
Comment (lukaszstolarczuk) [inline src/utils/utils_common.h:48]: nit: u could use `ALIGN_UP` here

Comment (PatKamin) [reply src/utils/utils_common.h]: Done

File: src/utils/utils_common.h [hunk 16]
```diff
@@ -40,6 +40,12 @@ typedef enum umf_purge_advise_t {
 #define IS_NOT_ALIGNED(value, align)                                           \
     ((align != 0 && (((value) & ((align)-1)) != 0)))
 #define ALIGN_UP(value, align) (((value) + (align)-1) & ~((align)-1))
+#ifndef ALIGN_UP_SAFE
```
Comment (lplewa) [inline src/utils/utils_common.h:43]: This ifndef looks suspicious. Do we need it. If yes why?

Comment (PatKamin) [reply src/utils/utils_common.h:43]: It was needed because of the doubled definition of this macro. I've removed the other definition now. Thanks for catching this one, less rubbish.

File: src/provider/provider_os_memory.c [hunk 17]
```diff
@@ -864,7 +898,12 @@ static membind_t membindNext(os_memory_provider_t *provider,
         membind.node++;
         membind.node %= provider->nodeset_len;
         membind.bitmap = provider->nodeset[membind.node];
-        membind.bind_size = ALIGN_UP(provider->part_size, membind.page_size);
+        membind.bind_size =
+            ALIGN_UP_SAFE(provider->part_size, membind.page_size);
+        if (membind.bind_size == 0) {
+            LOG_ERR("part_size is too big, page align failed");
+            return membind;
+        }
```
[the remark this answers is not in the document]
Comment (PatKamin) [reply src/provider/provider_os_memory.c]: You're right. Moved aligning of `size` up. Left `ALIGN_UP` macro for `part_size` as it can be equal to zero.

File: test/test_valgrind.sh [hunk 18]
```diff
@@ -117,6 +117,9 @@ for test in $(ls -1 umf_test-*); do
 	umf_test-memspace_lowest_latency)
 		FILTER='--gtest_filter="-*allocLocalMt*"'
 		;;
+	umf_test-memoryPool)
+		FILTER='--gtest_filter="-*allocMaxSize*"'
```
Comment (lplewa) [inline test/test_valgrind.sh:121]: Why we are disabling valgrind here - we are expecting fail of the allocation, so valgrind should be fine.

File: src/utils/utils_common.h (hunk 16 above)
Comment (PatKamin) [reply src/utils/utils_common.h]: We don't need that, my omission, now removed.

File: test/test_valgrind.sh (hunk 18 above)
Comment (PatKamin) [reply test/test_valgrind.sh:121]: Memcheck fails the test with SIZE_MAX allocation: https://github.com/PatKamin/unified-memory-framework/actions/runs/11481177678/job/31952726455#step:6:116

**Review by lplewa — APPROVED** [review]

File: src/base_alloc/base_alloc_linear.c [hunk 19]
```diff
@@ -122,15 +126,24 @@ void *umf_ba_linear_alloc(umf_ba_linear_pool_t *pool, size_t size) {
...
     if (pool->metadata.size_left < aligned_size) {
         size_t pool_size = MINIMUM_LINEAR_POOL_SIZE;
         size_t usable_size =
             pool_size - offsetof(umf_ba_next_linear_pool_t, data);
         if (usable_size < aligned_size) {
             pool_size += aligned_size - usable_size;
-            pool_size = ALIGN_UP(pool_size, ba_os_get_page_size());
+            pool_size = ALIGN_UP_SAFE(pool_size, ba_os_get_page_size());
+            if (pool_size == 0) {
+                utils_mutex_unlock(&pool->metadata.lock);
```
Comment (KFilipek) [inline src/base_alloc/base_alloc_linear.c:143]: According to the Helgrind log, I believe the problem is here. Are you sure it should be done here?

Comment (PatKamin) [reply src/base_alloc/base_alloc_linear.c:143]: Yes, I have to release the mutex before returning from this function.

**Comment by PatKamin** [comment]
Rebased
